controller/src/main/java/io/pravega/controller/store/stream/ZkInt96Counter.java

```diff
@@ -0,0 +1,151 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.controller.store.stream;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.pravega.common.lang.AtomicInt96;
+import io.pravega.common.lang.Int96;
+import lombok.extern.slf4j.Slf4j;
+
+import javax.annotation.concurrent.GuardedBy;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+
+@Slf4j
+public class ZkInt96Counter {
+    /**
+     * This constant defines the size of the block of counter values that will be used by this controller instance.
+     * The controller will try to get current counter value from zookeeper. It then tries to update the value in store
+     * by incrementing it by COUNTER_RANGE. If it is able to update the new value successfully, then this controller
+     * can safely use the block `previous-value-in-store + 1` to `previous-value-in-store + COUNTER_RANGE` No other controller
+     * will use this range for transaction id generation as it will be unique assigned to current controller.
+     * If controller crashes, all unused values go to waste. In worst case we may lose COUNTER_RANGE worth of values everytime
+     * a controller crashes.
+     */
+    @VisibleForTesting
+    static final int COUNTER_RANGE = 10000;
+    @VisibleForTesting
+    static final String COUNTER_PATH = "/counter";
+
+    private final Object lock;
+    @GuardedBy("lock")
+    private final AtomicInt96 limit;
+    @GuardedBy("lock")
+    private final AtomicInt96 counter;
+    @GuardedBy("lock")
+    private volatile CompletableFuture<Void> refreshFutureRef;
```

Comment:
There is no need for something guarded by a lock to be volatile. The lock must cover all places it is used and provides all the synchronization needed.